### ISSUE TITLE
Possible regression of F.scaled_dot_product_attention on CPU in PyTorch 2.5

### ISSUE THREAD

**Issue by BenjaminBossan** [pytorch/pytorch#141128]
### 🐛 Describe the bug

When running `F.scaled_dot_product_attention` with an input matrix that contains NaNs on CPU, with PyTorch 2.4, the output is a NaN matrix, but with PyTorch 2.5, it is a zeros matrix.

```python
import contextlib
import torch
import torch.nn.functional as F
import torch.nn as nn

device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")

contexts = {
    "no context": contextlib.nullcontext(),
}
if device.type == "cpu":
    contexts["SDPBackend.FLASH_ATTENTION"] = nn.attention.sdpa_kernel(nn.attention.SDPBackend.FLASH_ATTENTION)
    contexts["SDPBackend.MATH"] = nn.attention.sdpa_kernel(nn.attention.SDPBackend.MATH)

with torch.device(device):
    for name, ctx in contexts.items():
        with ctx:
            query = torch.randn(2, 4, 16, 8)
            key = torch.randn(2, 4, 16, 8)
            value = torch.randn(2, 4, 16, 8)

            hidden_states = F.scaled_dot_product_attention(
                query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False
            )
            assert torch.isfinite(hidden_states).all()

            query += torch.nan
            hidden_states = F.scaled_dot_product_attention(
                query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False
            )
            try:
                assert torch.isnan(hidden_states).all()
                print(f"Succeeded with device: {device}, torch version: {torch.__version__}, context: {name}")
            except AssertionError:
                print(f"Failed    with device: {device}, torch version: {torch.__version__}, context: {name}")
```

I get the following outputs:

```
# with CPU, torch 2.4.1
$ CUDA_VISIBLE_DEVICES='' python foo.py
Succeeded with device: cpu, torch version: 2.4.1+cu121, context: no context
Succeeded with device: cpu, torch version: 2.4.1+cu121, context: SDPBackend.FLASH_ATTENTION
Succeeded with device: cpu, torch version: 2.4.1+cu121, context: SDPBackend.MATH

# with CPU, torch 2.5.1
$ CUDA_VISIBLE_DEVICES='' python foo.py
Failed    with device: cpu, torch version: 2.5.1+cu124, context: no context
Failed    with device: cpu, torch version: 2.5.1+cu124, context: SDPBackend.FLASH_ATTENTION
Succeeded with device: cpu, torch version: 2.5.1+cu124, context: SDPBackend.MATH

# with CUDA, torch 2.4.1:
$ CUDA_VISIBLE_DEVICES=0 python foo.py
Succeeded with device: cuda:0, torch version: 2.5.1+cu124, context: no context

# with CUDA, torch 2.5.1:
$ CUDA_VISIBLE_DEVICES=0 python foo.py
Succeeded with device: cuda:0, torch version: 2.4.1+cu121, context: no context
```

The MATH backend seems to fix the issue for PyTorch 2.5.

### Versions

Python 3.11.9, torch versions as indicated.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki

**Comment by drisspg** [comment]
This is expected, and IMO kind of edge casey. We recently made updates to sdpa support fully masked out rows here:https://github.com/pytorch/pytorch/pull/131060
and I imagine this is the cause

**Comment by BenjaminBossan** [comment]
Thanks for the response. Just to be clear, it is intended that `scaled_dot_product_attention` will give different outputs for CPU and CUDA?

**Comment by drisspg** [comment]
Ooops I might have misread let me take another look

**Comment by drisspg** [comment]
Ohh you are right this doesn't seem right. and is isolated to the FlashAttention on CPU Impl. I have a fix here: https://github.com/pytorch/pytorch/pull/141519

**Comment by sayakpaul** [comment]
@drisspg thank you! Could you give us a heads-up when this has been patched. Will update our CI accordingly.
